File: Orbs2/src/main/java/world/World.java
package world;

import javafx.scene.image.Image;
import javafx.scene.canvas.GraphicsContext;
import javafx.scene.input.KeyCode;
import javafx.scene.layout.TilePane;
import java.util.*;

import je.files.ReadFile;
import je.collections.ArrayConversion;
import je.visual.Vector2D;

import states.WorldState;
import main_package.Assets;
import entities.*;

public class World extends Entity {

	/********************
	*					*
	*	  VARIABLES		*
	*					*
	*********************/

	// A reference to the world state.
   	WorldState worldState;

   	// The name of the file containing the map data.
   	String mapName;

   	// The size of the map
   	int mapSize;

   	// The matrix of tiles.
   	ArrayList<Tile> tiles;

   	// Whether or not the tiles are moving.
   	boolean moving;


   	// A reference to the player that is currently in this world.
   	Player player;

   	// All of the entities in the game world.
   	ArrayList<Entity> entities;




	/********************
	*					*
	*   INITIALIZATION	*
	*					*
	*********************/

	public World(Player player, WorldState ws, String mapName, int size) {
		super(new Vector2D(), ws);
		this.worldState = ws;
		this.mapName = mapName;
		this.mapSize = size;
		this.moving = false;
		this.player = player;
		this.entities = new ArrayList<Entity>();

		this.configureMap();
	}


	/********************
	*					*
	*	   SETTERS		*
	*					*
	*********************/

	// Creates the matrix of id's for the map.
	private void configureMap() {
		ReadFile reader = new ReadFile("/Users/adeolauthman/Documents/AdeolasCodingStuff/JavaPrograms/Orbs2/resources/maps/"+this.mapName+".txt");
		try {
			String s = reader.read();
			String[] arr = s.split("-");
			
			String[][] map2D = (String[][])ArrayConversion.OneToTwo(arr, this.mapSize);

			// Start making the tiles and the map.
			this.tiles = new ArrayList<Tile>();
			this.makeMap(map2D);

		} catch(Exception er) { er.printStackTrace(); }
	}
	

	// Handles actually making the tile map.
	private void makeMap(String[][] map) {
		for(int i = 0; i < map.length; i++) {
			for(int j = 0; j < map[0].length; j++) {
				
				switch(map[i][j]) {
					case "0": this.createTile(j, i, Assets.GRASS_1); break;
					case "20": this.createTile(j, i, Assets.TREE_TOP); break;
					case "40": this.createTile(j, i, Assets.TREE_BOTTOM); break;
					default: this.createTile(j, i, Assets.GRASS_1); break;
				}

			}
		}
	}


	// Creates a tile.
	private void createTile(int i, int j, Image image) {
		Tile tile = new Tile(new Vector2D(i,j), image, this.worldState);
		tiles.add(tile);
	}


	// Handles the key actions
	public void keyActions(KeyCode code) {
		if(code == KeyCode.UP) {
			player.changeDirection(Direction.UP);
			this.tiles.forEach(e -> { e.yVel = this.speed; });
		}
		if(code == KeyCode.DOWN) {
			player.changeDirection(Direction.DOWN);
			this.tiles.forEach(e -> { e.yVel = -this.speed; });
		}
		if(code == KeyCode.LEFT) {
			player.changeDirection(Direction.LEFT);
			this.tiles.forEach(e -> { e.xVel = this.speed; });
		}
		if(code == KeyCode.RIGHT) {
			player.changeDirection(Direction.RIGHT);
			this.tiles.forEach(e -> { e.xVel = -this.speed; });
		}
	}


	// Handles key releases
	public void releaseKeyActions(KeyCode code) {
		if(code == KeyCode.UP) {
			this.tiles.forEach(e -> { e.yVel = 0; });
		}
		if(code == KeyCode.DOWN) {
			this.tiles.forEach(e -> { e.yVel = 0; });
		}
		if(code == KeyCode.LEFT) {
			this.tiles.forEach(e -> { e.xVel = 0; });
		}
		if(code == KeyCode.RIGHT) {
			this.tiles.forEach(e -> { e.xVel = 0; });
		}
	}


	/********************
	*					*
	*	   GETTERS		*
	*					*
	*********************/

	public Vector2D getPosition() {
		return position;
	}

	public ArrayList<Entity> getEntities() {
		return entities;
	}

	public boolean isMoving() {
		return moving;
	}


	/********************
	*					*
	*	   ABSTRACT		*
	*					*
	*********************/

	public void initialize() {
		if(player != null) { player.initialize(); }
		if(tiles != null) {
			this.tiles.forEach(e -> { e.initialize(); });
		}
	}

	public void update() {
		if(player != null) { player.update(); }
		if(tiles != null) {
			this.tiles.forEach(e -> { e.update(); });

			if(this.tiles.get(0).xVel == 0 && this.tiles.get(0).yVel == 0) {
				this.moving = false;
			} else {
				this.moving = true;
			}
		}
	}

	public void draw() {
		if(tiles != null) {
			this.tiles.forEach(e -> { e.draw(); });
		}
		if(player != null) { player.draw(); }
	}

} // End of class.

File: Orbs2/src/world/World.kt
package world

import entities.Entity
import entities.Player
import javafx.scene.input.KeyCode
import je.collections.ArrayConversion
import je.files.ReadFile
import je.visual.Vector2D
import main_package.drawCollisionBoxes
import main_package.printTilesRendered
import states.WorldState
import tiles.BlueRug_BL
import tiles.BlueRug_BR
import tiles.BlueRug_TL
import tiles.BlueRug_TR
import tiles.GrassTile
import tiles.HouseOne_BL
import tiles.HouseOne_BR
import tiles.HouseOne_BaseL
import tiles.HouseOne_BaseR
import tiles.HouseOne_DoorL
import tiles.HouseOne_DoorR
import tiles.HouseOne_DoorTL
import tiles.HouseOne_DoorTR
import tiles.HouseOne_Inner
import tiles.HouseOne_L
import tiles.HouseOne_R
import tiles.HouseOne_TL
import tiles.HouseOne_TR
import tiles.HouseOne_Top
import tiles.HouseOne_Wall
import tiles.RedRug_BL
import tiles.RedRug_BR
import tiles.RedRug_TL
import tiles.RedRug_TR
import tiles.Tile
import tiles.TreeBottom
import tiles.TreeTop
import tiles.Well
import tiles.WoodFloor
import java.util.ArrayList
import HouseOne_InnerSL


public open class World(player: Player?, ws: WorldState?, mapName: String?, size: Int) {
	
	/********************
	 *					*
	 *	   VARIABLES	*
	 *					*
	 ********************/
	
	// A reference to the world state.
	var worldState: WorldState? = null
	
	// The rendering camera.
	var camera: Camera? = null
	
	// The name of the file containing the map data.
	var mapName: String? = null
	
	// The size of the map
	var mapSize: Int = 0
	
	// The speed of the world (same thing as in the Entity class)
	var speed: Float = 0.15f
	
	// The matrix of tiles.
	var tilesL1: ArrayList<Tile>? = null
	var tilesL2: ArrayList<Tile>? = null
	
	
	// A reference to the player that is currently in this world.
	var player: Player? = null
	
	// All of the entities in the game world.
	var entities: ArrayList<Entity>? = null

	
	
	init {
		this.worldState = ws
		this.camera = Camera(0f, 0f)
		this.mapName = mapName
		this.mapSize = size
		this.player = player
		this.entities = ArrayList<Entity>()
		this.configureMap()
	}

	/********************
	 *					*
	 *	   SETTERS		*
	 *					*
	 ********************/
	
	// Creates the matrix of id's for the map.
	private fun configureMap() {
		var reader = ReadFile(World::class.java.getResource("../resources/maps/${mapName}_L1.txt").getFile())
		try {
			val s = reader.read().toString()
			val arr = s.split(",").toTypedArray()
			val map2D = ArrayConversion.OneToTwo(arr, this.mapSize)
			
			// Start making the tiles and the map.
			this.tilesL1 = ArrayList<Tile>()
			this.makeMap(map2D, tilesL1!!)
			
			reader = ReadFile(World::class.java.getResource("../resources/maps/${mapName}_L2.txt").getFile())
			val s2 = reader.read().toString()
			val arr2 = s2.split(",").toTypedArray()
			val map2D2 = ArrayConversion.OneToTwo(arr2, this.mapSize)
			
			this.tilesL2 = ArrayList<Tile>()
			this.makeMap(map2D2, tilesL2!!)
		} catch (er: Exception) {
			er.printStackTrace()
		}
	}

	
	// Handles actually making the tile map.
	private fun makeMap(map: Array<Array<String>>?, tiles: ArrayList<Tile>) {
		for (i in map!!.indices) {
			for (j in 0..map[0].size - 1) {
				when (map[i][j]) {
					"1" -> tiles.add( GrassTile(Vector2D(j.toFloat(), i.toFloat()), worldState) );
					"22","68","88","108" -> tiles.add( WoodFloor(Vector2D(j.toFloat(), i.toFloat()), worldState) );
					"3" -> tiles.add( Well(Vector2D(j.toFloat(), i.toFloat()), worldState) );
					"21" -> tiles.add( TreeTop(Vector2D(j.toFloat(), i.toFloat()), worldState) );
					"41" -> tiles.add( TreeBottom(Vector2D(j.toFloat(), i.toFloat()), worldState) );
					"47" -> tiles.add( HouseOne_TL(Vector2D(j.toFloat(), i.toFloat()), worldState) );
					"67" -> tiles.add( HouseOne_L(Vector2D(j.toFloat(), i.toFloat()), worldState) );
					"87" -> tiles.add( HouseOne_BL(Vector2D(j.toFloat(), i.toFloat()), worldState) );
					"107" -> tiles.add( HouseOne_BaseL(Vector2D(j.toFloat(), i.toFloat()), worldState) );
					"110" -> tiles.add( HouseOne_Wall(Vector2D(j.toFloat(), i.toFloat()), worldState) );
					"111", "112" -> tiles.add( HouseOne_DoorL(Vector2D(j.toFloat(), i.toFloat()), worldState) );
					"113" -> tiles.add( HouseOne_DoorR(Vector2D(j.toFloat(), i.toFloat()), worldState) );
					"49" -> tiles.add( HouseOne_TR(Vector2D(j.toFloat(), i.toFloat()), worldState) );
					"69" -> tiles.add( HouseOne_R(Vector2D(j.toFloat(), i.toFloat()), worldState) );
					"89" -> tiles.add( HouseOne_BR(Vector2D(j.toFloat(), i.toFloat()), worldState) );
					"90", "91" -> tiles.add( HouseOne_Inner(Vector2D(j.toFloat(), i.toFloat()), worldState) );
					"92" -> tiles.add( HouseOne_DoorTL(Vector2D(j.toFloat(), i.toFloat()), worldState) );
					"93" -> tiles.add( HouseOne_DoorTR(Vector2D(j.toFloat(), i.toFloat()), worldState) );
					"109" -> tiles.add( HouseOne_BaseR(Vector2D(j.toFloat(), i.toFloat()), worldState) );
					"23" -> tiles.add( BlueRug_TL(Vector2D(j.toFloat(), i.toFloat()), worldState) );
					"24" -> tiles.add( BlueRug_TR(Vector2D(j.toFloat(), i.toFloat()), worldState) );
					"43" -> tiles.add( BlueRug_BL(Vector2D(j.toFloat(), i.toFloat()), worldState) );
					"44" -> tiles.add( BlueRug_BR(Vector2D(j.toFloat(), i.toFloat()), worldState) );
					"48" -> tiles.add( HouseOne_Top(Vector2D(j.toFloat(), i.toFloat()), worldState) );
					"25" -> tiles.add( RedRug_TL(Vector2D(j.toFloat(), i.toFloat()), worldState) );
					"26" -> tiles.add( RedRug_TR(Vector2D(j.toFloat(), i.toFloat()), worldState) );
					"45" -> tiles.add( RedRug_BL(Vector2D(j.toFloat(), i.toFloat()), worldState) );
					"46" -> tiles.add( RedRug_BR(Vector2D(j.toFloat(), i.toFloat()), worldState) );
					"52" -> tiles.add( HouseOne_InnerSL(Vector2D(j.toFloat(), i.toFloat()), worldState) );
				}
			}
		}
	}


	
	/********************
	 *					*
	 *	   GETTERS		*
	 *					*
	 ********************/

	
	
	
	
	/********************
	 *					*
	 *	   ABSTRACT		*
	 *					*
	 ********************/
	fun initialize() {
		if (player != null) {
			player!!.initialize()
		}
		
		/* the Initialize methods in Tile and Entity don't actually do anything, so don't waste
		   time calling them. */ 
//		if (tiles != null) {
//			this.tiles!!.forEach({ e -> e.initialize() })
//		}
	}

	fun update() {
		if (player != null) {
			player!!.update()
			camera!!.update(player!!)
		}
		
		
		/* For right now, you don't need to update each tile. */
//		if (tiles != null) {
//			this.tiles!!.forEach({ e -> e.update() })
//		}
	}

	fun draw() {
		worldState!!.graphics.translate(-camera!!.position.X.toDouble(), -camera!!.position.Y.toDouble())
		
		
		if (tilesL1 != null && tilesL2 != null) {
			// Only draw the tiles that are within the camera view
			this.tilesL1!!
				.filter { it -> camera!!.touching(it); }
				.forEach { e -> e.draw(); }
			this.tilesL2!!
				.filter { it -> camera!!.touching(it); }
				.forEach{ e -> e.draw() }
		}
		if (player != null) {
			player!!.draw()
		}
		
		worldState!!.graphics.translate(camera!!.position.X.toDouble(), camera!!.position.Y.toDouble())
	}
	
	
	
	/********************
	 *					*
	 *	    DEBUG		*
	 *					*
	 ********************/
	
	fun debug(code: KeyCode) {
		if(code == KeyCode.R) {
			printTilesRendered()
		}
		if(code == KeyCode.D) {
			drawCollisionBoxes()
		}
	}

} // End of class.